Handle always included [de]activation scripts

jakirkham opened the pull request:
Tries to handle the now always included [de]activation scripts.

cc @pelson @bollwyvl @msarahan

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -15,8 +15,14 @@ build:
+    - ssl/cert.pem                                                       # [unix]
+    - ssl/cacert.pem                                                     # [unix]
+    - Library\\ssl\\cert.pem                                             # [win]
+    - Library\\ssl\\cacert.pem                                           # [win]
```
I really don't understand how this still isn't needed for Windows, but it appears it still isn't.

msarahan commented:
No library in win path, just prefix:

https://github.com/conda/conda/blob/master/bin/activate.bat#L76

jakirkham commented:
> I won't pretend to fully understand it, but it has gone green!

It is fortunately not that bad. As I mentioned to @msarahan earlier, the problem is we need to use `python` and `certifi` as build dependencies. So, we have this chain `certifi` -> `python` -> `openssl` -> `ca-certificates`. As a result, we find ourselves depending on the last build of `ca-certificates` when we build the next one.

I've added some tests to make sure it works. Though I think it is my bad not adding them before. 😞

> If you bump the build number, I don't have any objections 👍

I can do that, but only because I deleted the old build number. If I didn't do that, we would be in the jackpot w.r.t. `always_include_files` as we would be missing the old [de]activate scripts. Though @msarahan is softening that constraint so it doesn't fail to build if they are missing. So ideally we won't have to think about that problem much longer.

jakirkham commented:
> No library in win path, just prefix:

> https://github.com/conda/conda/blob/master/bin/activate.bat#L76

Ok, we can fix that. Though this is a holdover from the PR that added the activation scripts. See this [line]( https://github.com/conda-forge/ca-certificates-feedstock/pull/2/files#diff-d04c86b6bb20341f5f7c53165501a393R17 ).

In any event, it still doesn't explain why `always_include_files` doesn't pick them up. They are present in that Windows package.

jakirkham commented:
I think I have addressed the new round of comments. Have tested this locally with `conda` 4.1.6 on my Mac and can verify it works. Can't say anything about Windows though.

jakirkham commented:
Removing WIP as this is done on my end. Please let me know if needs anything else @msarahan or @pelson.

jakirkham commented on recipe/activate.bat:
```diff
@@ -1,2 +1,3 @@
-set "SSL_CERT_DIR=%CONDA_ENV_PATH%\ssl"
+:: handle conda...   >=4.1.3        <= 4.1.2
+set "SSL_CERT_DIR=%CONDA_PREFIX%%CONDA_ENV_PATH%\ssl"
```
Maybe this works in Windows? If not this, we may need to do something else.

jakirkham replied:
This doesn't have any `export` statements AFAICT.

msarahan commented on recipe/meta.yaml:
```diff
@@ -7,16 +7,22 @@ package:
+   #- Library\\ssl\\cacert.pem                                           # [win]
+    - etc/conda/activate.d/ca-certificates_activate.sh                   # [unix]
+    - etc/conda/deactivate.d/ca-certificates_deactivate.sh               # [unix]
+   #- etc\\conda\\activate.d\\ca-certificates_activate.bat               # [win]
```
Can we re-enable these?

msarahan replied:
I guess you're waiting on the conda-build update that allows your build to succeed.  It'll be up soon.

jakirkham replied:
Yep. Now it's switched though.

jakirkham commented:
This is a pretty weird error on AppVeyor. See the snippet below from this [log]( https://ci.appveyor.com/project/conda-forge/ca-certificates-feedstock/build/1.0.15/job/41qwxmblr1wfqx5u#L282 ). It's like it is using the wrong activate script. However, I don't think we are including it.

```
C:\conda\conda-bld\test-tmp_dir>activate _test 
The system cannot find the file specified.
'export' is not recognized as an internal or external command,
operable program or batch file.
'export' is not recognized as an internal or external command,
operable program or batch file.
The system cannot find the path specified.
TESTS FAILED: ca-certificates-2016.2.28-3
Command exited with code 1
```

jakirkham commented:
This happened with `conda-build` version 1.21.5 and 1.21.4 (though the activation syntax changed slightly) as evidenced by this [log]( https://ci.appveyor.com/project/conda-forge/ca-certificates-feedstock/build/1.0.15/job/41qwxmblr1wfqx5u#L282 ). Not sure what I'm doing wrong, but it points to the wrong activation script being copied.

jakirkham commented on recipe/bld.bat:
```diff
@@ -10,9 +10,9 @@ if errorlevel 1 exit 1
-    IF NOT EXIST %LIBRARY_PREFIX%\etc\conda\%%F.d MKDIR %LIBRARY_PREFIX%\etc\conda\%%F.d
-    COPY %RECIPE_DIR%/%%F.bat %LIBRARY_PREFIX%\etc\conda\%%F.d\%PKG_NAME%_%%F.bat
+    IF NOT EXIST %PREFIX%\etc\conda\%%F.d MKDIR %PREFIX%\etc\conda\%%F.d
+    COPY %RECIPE_DIR%/%%F.bat %PREFIX%\etc\conda\%%F.d\%PKG_NAME%_%%F.bat
```
Looks like we are copying a `.bat` file to me. 😕

jakirkham commented:
The last successful build used at `conda-build` 1.21.4 too.

msarahan commented:
yeah, not your problem.  I think the code in conda-build needs to include the extension (or not, as appropriate.)  Will put this into another 1.21 release.

jakirkham commented:
Maybe there is a `conda-build` bug? IDK.

The only difference I see between the last working and first failing build are these [changes]( https://github.com/conda-forge/ca-certificates-feedstock/compare/cb4fe08e7f4df9102c889ab0be75145b90c1b9e8...5aac6d25fc4b2f84d2e433ca024d975a7b4836a1 ). The versions of everything between this [passing build]( https://ci.appveyor.com/project/conda-forge/ca-certificates-feedstock/build/1.0.12/job/wsx4b9jq5wli1vqb ) and this [failing build]( https://ci.appveyor.com/project/conda-forge/ca-certificates-feedstock/build/1.0.14/job/kki5gtywi3ffi28p ) look identical to me.

jakirkham commented:
Alright, there is one subtle difference I do see in dependencies that I missed before. We have a build number bump (`0` -> `1`) in `conda-forge`'s `python` version 3.5.2. Not sure how that affects anything, but it seems worth noting.

jakirkham commented:
So it sounds like we are waiting for PR ( https://github.com/conda/conda-build/pull/1107 ) to make it into a release. Though maybe we should test it here to make sure it fixes the issue.

Is there anything we can do to get this in for the meantime? If not, we should pull build number 1 as that causes issues in some cases due to the very problem that @bollwyvl was trying to fix.

183amir commented:
A new conda build was just released.

jakirkham commented:
Restarted AppVeyor. Let's see what happens.

msarahan commented:
Dang, this queue is pretty crazy.  I'm itching to know if the new Conda Build fixes things...

jakirkham commented:
😭 

```
The system cannot find the file specified.
'export' is not recognized as an internal or external command,
operable program or batch file.
'export' is not recognized as an internal or external command,
operable program or batch file.
The system cannot find the path specified.
```

jakirkham commented:
I have labeled build number 1 as broken so it shouldn't interfere any more. Sorry @183amir. Seems we need a bit more time to figure this out.

msarahan commented:
Rats.  I'll investigate this more tomorrow.  I'm surprised that this didn't work.

jakirkham commented:
No worries. This activation stuff (particularly on Windows) seems to be a tricky business. I'm glad to see `conda` and `conda-build` getting smarter about it though.

jakirkham commented:
So the main motivation (for me at least) to pursue this is gone. Therefore am closing this out.
